### DISCUSSION TITLE
docs: network.WLAN in STA_IF mode: which parameters are persistent?

### DISCUSSION THREAD

**Discussion by massimosala** [micropython/micropython#11548] [ESP8266]
I hit the problem on esp8266, but it is primarly a lack of documentation.

Example: this basic code to init the micro as Wifi client:

```
network.country("it")
WIFI = network.WLAN(network.STA_IF)
WIFI.config(dhcp_hostname = "test")
WIFI.connect(ssid, pwd)
```

From the docs I read

> ESP8266 automatically reconnects to the last Access Point when STA_IF is active, even after reboot/reset

Indeed after a  reboot the ESP will auto-reconnect the the access point.
But
- It has lost the hostname: in the router I see the factory default "ESP_xxx" (where xxx seems to be the machine unique id).

- I don't know about the country setting, if it is persistent between reboots.
I haven't the radio equiment to check the esp radio emissions, regarding WiFi compliance to national requirements.

Questions:
1) Is it confirmed the esp8266 store only the SSID and password ?
2) Is this a  behaviour of the esp8266 of also on other micros, like esp32 ?
3) At power up, is it possible to mantain the working WiFi connection and set the hostname?
4) which country codes are supported by _network.country()_ ?

I cannot find hints.
Please anybody point me to the docs -  - or how can we improve the existing ones.

**Comment by glenn20** [comment]
This behaviour on the esp8266 is entirely a function of the Espressif API, and the best source of information **should** be the Espressif docs for the ESP8266 (nonOS SDK) (eg. google "esp8266 nonos sdk pdf"). However, they can be ambiguous and incomplete. It's just not feasible (or appropriate) to replicate all that documentation into micropython docs.

On esp8266, the espressif SDK saves the wifi parameters (at least ssid and password into non-volatile RAM and restores then on reboot). Working out exactly which parameters it saves is, unfortunately an exercise for the reader ;). The answer *may* be in the docs (check the sdk pdf).

PR #9460 adds an option to disable the auto-reconnect "feature" on the esp8266.

Personally, I detest this auto-connect misfeature of the esp8266. I always run `sta.disconnect()` on reboot and then manually connect. Then you can configure whichever options are important to you.

The esp32 does not have this "feature" enabled by default - but it is possible to configure simlar behaviour using the ESP-IDF, but that is not supported by micropython (thankfully ;-)).

**Comment by glenn20** [comment]
Oh, I nearly forgot.... re. Question 4... the country setting is not currently used for micropython on the esp8266. In fact, it is only used currently for the cyw43 wifi driver (eg. the rpi pico-w).

The allowed values __should__ be any valid two-letter "wifi country codes" (just google that phrase for a complete list), however some devices support only a limited list.

**Reply by massimosala** [reply]
> Oh, I nearly forgot.... re. Question 4... the country setting is not currently used for micropython on the esp8266. In fact, it is only used currently for the cyw43 wifi driver (eg. the rpi pico-w).
> 
> The allowed values **should** be any valid two-letter "wifi country codes" (just google that phrase for a complete list), however some devices support only a limited list.

Hi @glenn20 
Thank you for the  details.

I propose to **take away _network.country()_ from all the ports where it isn't supported**, and update the official docs.

---

> On esp8266, the espressif SDK saves the wifi parameters (at least ssid and password into non-volatile RAM and restores then on reboot). Working out exactly which parameters it saves is, unfortunately an exercise for the reader ;). The answer may be in the docs (check the sdk pdf).

About _dhcp_hostname_: I need to do more tests to see if it's persistent or not.
If anyone has any clues, please comment here.
I'm keeping this issue open for now, maybe I can find more details.

**Comment by Tobias4git** [comment]
I can confirm the (at least for me) undesired behavior: The hostname is not persistent. After a reset the hostname in the dhcp request is ESP-XXYYZZ.
I think the ESP re-connects automatically. Whenever the python interpreter wants to change it, it's too late. 

Funny enough even the default hostname gets ignored.
If you do a hard reset without a main.py, then:
import network
WIFI = network.WLAN(network.STA_IF)
WIFI.config('dhcp_hostname')
=> 'mpy-esp8266'
the same for 
network.hostname()
=> 'mpy-esp8266'
You would think that if you continue to connect, the router would assume you're now mpy-esp8266....
No.
You still get ESP-XXYYZZ.

I found an ugly workaround:
Open the firmware ESP8266_GENERIC-YYYYMMDD....bin with your preferred editor.
Search for the string ESP_%02X%02X%02X  (warning: it's an underscore!)
Replace it with the required hostname, consume or add null bytes if needed. The size of the file must not change.
Flash it.
Brrrrr.

**Reply by Josverl** [reply]
Ugly workaround indeed.

I assume the DHCP server registered the first name ESPxxxx with the (m)dns service, and that the ttl for that record is longer than the time between the first and second DHCP request. 
To be sure your need to sniff the trafic on the wifi/router interface
